certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -616,26 +596,48 @@ def _enable_redirect(self, domain, unused_options):
         vhost = None
         # If there are blocks listening plaintextishly on self.DEFAULT_LISTEN_PORT,
         # choose the most name-matching one.
+
+        # TODO: should we be adding redirects to all matching blocks?
+
         vhost = self.choose_redirect_vhost(domain, port)
 
         if vhost is None:
             logger.info("No matching insecure server blocks listening on port %s found.",
                 self.DEFAULT_LISTEN_PORT)
+            return
+
+        if vhost.ssl:
+            new_vhost = self.parser.duplicate_vhost(vhost,
+                only_directives=['listen', 'server_name'])
+
+            def _ssl_match_func(directive):
+                return 'ssl' in directive
+
+            def _no_ssl_match_func(directive):
+                return 'ssl' not in directive
+
+            # remove all ssl addresses from the new block
+            self.parser.remove_server_directives(new_vhost, 'listen', match_func=_ssl_match_func)
+
+            # remove all non-ssl addresses from the existing block
+            self.parser.remove_server_directives(vhost, 'listen', match_func=_no_ssl_match_func)
+
+            vhost = new_vhost
+
+        if self._has_certbot_redirect(vhost):
```

Comment:
This is no longer true if we go with my comments above, but if we don't, I think this can and should be `elif` for increased clarity here. Since the vhost can't possibly have an existing redirect with only `listen` and `server_name` directives, neither of these `if` statements should ever be true.